### PR TITLE
[WIP]Infinity Class

### PR DESCRIPTION
**Pull request by CodeMaxx**
- [x] Inherit Infinity from Number
- [x] Implement various member functions
- [x] Implement constants `Inf`, `NegInf`, `ComplexInf`
- [x] Differentiation
- [x] Printing
- [x] Added tests
- [ ] `pow`, `powr` and `div` in `Infinity` class
- [ ] Handle Infinity in Add, Mul, Pow class
- [ ] Add leftout tests
- [ ] Change name to `Infinity`
- [ ] Check for Infinity TODOs in `functions` like hyperbolic functions, log(0) etc.
- [ ] Implement for all directions if it seems necesary.
- [ ] Think of other possible enhancements

The class name `Infinity` was giving error on my machine which I possible believe is due to overlap with some already define `Infinity`. I'll look into this matter later. Currently the class is name `Infinit`.

@isuruf @certik Please Review and comment.

### CODE REVIEW

**Comment by CodeMaxx** [comment]
I will not be available in the following week due to my final exams. Will be back by Saturday.

Cheers!

File: symengine/derivative.cpp [hunk 1]
```diff
@@ -519,6 +520,12 @@ class DiffImplementation
         return self.diff_impl(x);
     }
 
+    static RCP<const Basic> diff(const Infinit &self,
```
Comment (isuruf) [inline symengine/derivative.cpp]: This is not necessary

File: symengine/infinity.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                                     "encountered");
+        else
+            throw std::runtime_error("Indeterminate Expression: `infinity +- "
+                                     "infinity` encountered. Directions don't "
+                                     "match");
+    } else
+        return make_rcp<const Infinit>(*this);
+}
+
+RCP<const Number> Infinit::sub(const Number &other) const
```
Comment (isuruf) [inline symengine/infinity.cpp]: This is not necessary.

File: symengine/infinity.cpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+bool Infinit::is_negative_infinity() const
+{
+    return _direction->is_negative();
+}
+
+RCP<const Number> Infinit::add(const Number &other) const
+{
+    if (not is_a<Infinit>(other))
+        return make_rcp<const Infinit>(*this);
```
Comment (isuruf) [inline symengine/infinity.cpp]: `return this->rcp_from_this();`

**Comment by isuruf** [comment]
Looks good to me so far

File: symengine/derivative.cpp (hunk 1 above)
Comment (CodeMaxx) [inline symengine/derivative.cpp]: I see. We already have it defined for `Number`. Will remove this.

File: symengine/infinity.cpp (hunk 2 above)
Comment (CodeMaxx) [inline symengine/infinity.cpp]: You mean we can use `add` with a negative number instead ?

File: symengine/infinity.cpp (hunk 3 above)
Comment (CodeMaxx) [inline symengine/infinity.cpp]: Are these different or is it SymEngine's convention?

**Comment by CodeMaxx** [comment]
@isuruf Anything you would like added to the TODO list in the PR comment?

**Comment by nishnik** [comment]
@CodeMaxx Why don't we make it a singleton class ?

File: symengine/infinity.cpp (hunk 3 above)
Comment (CodeMaxx) [inline symengine/infinity.cpp]: @isuruf Ping

Comment (isuruf) [inline symengine/infinity.cpp]: Different. First one creates a new object, while the other returns a RCP to this.

File: symengine/infinity.cpp (hunk 2 above)
Comment (isuruf) [inline symengine/infinity.cpp]: No, this is the default implementation in `Number`. So this is redundant

File: symengine/infinity.cpp (hunk 3 above)
Comment (CodeMaxx) [inline symengine/infinity.cpp]: Cool got it.

File: symengine/infinity.cpp (hunk 2 above)
Comment (CodeMaxx) [inline symengine/infinity.cpp]: 👍

**Comment by CodeMaxx** [comment]
Note that I've closed this temporarily... I'll be continuing this in some time.

**Comment by isuruf** [comment]
Closed in favor of #1048

**Pull request closed without merging**